narciuch/assignment4/src/IslandFinder.java

```diff
@@ -0,0 +1,132 @@
+      return 0;
+    }
+    List<List<Integer>> islands = new ArrayList<>();
+    for (int row = 0; row < nrRows; row++) {
+      for (int column = 0; column < nrColumns; column++) {
+        if (!lands[row][column]) {
+          //Water tile, so we move to the next tile
+          continue;
+        }
+        List<Integer> currentTileIslandNrs = new ArrayList<>();
+        //Check above tile
+        if (row != 0) {
+          if (lands[row - 1][column]) {
+            int islandIndex = getIslandIndex(islands, getUniquePos(row - 1, column, nrColumns));
+            //Add above tile island nr to the list of islands the current tile is a part of
+            currentTileIslandNrs.add(islandIndex);
+          }
+        }
+        //Check in left tile
+        if (column != 0) {
+          if (lands[row][column - 1]) {
+            int islandIndex = getIslandIndex(islands, getUniquePos(row, column - 1, nrColumns));
+            //Add left tile island nr to the list of islands the current tile is a part of
+            currentTileIslandNrs.add(islandIndex);
+          }
+        }
+
+        //We have the list of islands the tile currently belongs to
+        //Now we update the list of islands to include this tile
+        int currentTile = getUniquePos(row, column, nrColumns);
+        if (currentTileIslandNrs.size() == 0) {
+          List<Integer> newIsland = new ArrayList<>();
+          newIsland.add(currentTile);
+          islands.add(newIsland);
+        } else if (currentTileIslandNrs.size() == 1) {
+          islands.get(currentTileIslandNrs.get(0)).add(currentTile);
+        } else {//Tile is a member of 2 islands
+          //Add tile arbitrarily to 1st island
+          islands.get(currentTileIslandNrs.get(0)).add(currentTile);
+          if (currentTileIslandNrs.get(0) != currentTileIslandNrs.get(1)) {
+            //Merge the 2 different islands into 1
+            islands = mergeIslands(islands, currentTileIslandNrs);
+          }
+        }
+      }
+    }
+
+    return islands.size();
+  }
+
+  /**
+   * Merges the inner arrays of islands, at the indices specified by the islandIndexes, into one
+   * array that it keeps in the islands array.
+   *
+   * <p>Returns a new copy of islands after this merge.
+   *
+   * @param islands       a list of integer lists, where each inner integer list represents an island
+   * @param islandIndexes the indices of those inner arrays in 'islands' that are to be merged
+   * @return a new copy of islands which has merged the islands as specified and not changed the
+   * other islands within the 'islands' nested list
+   */
+  private List<List<Integer>> mergeIslands(List<List<Integer>> islands, List<Integer> islandIndexes) {
+    List<Integer> mergeInto = islands.get(islandIndexes.get(0));
+    for (int mergeIslandIndex = 1; mergeIslandIndex < islandIndexes.size(); mergeIslandIndex++) {
+      int islandNr = islandIndexes.get(mergeIslandIndex);
+      for (int tileNr : islands.get(islandNr)) {
```

Comment:
Why not just `mergeInto.addAll(islands.get(islandNr);`?

Reply:
Yes, much simpler.